src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathFileGeneratorAuditListener.java

```diff
@@ -0,0 +1,139 @@
+ * See issue #102 https://github.com/checkstyle/checkstyle/issues/102
+ *
+ * @author Timur Tibeyev.
+ */
+public class XpathFileGeneratorAuditListener extends AutomaticBean implements AuditListener {
+
+    /** The " quote character. */
+    private static final String QUOTE_CHAR = "\"";
+
+    /**
+     * Helper writer that allows easy encoding and printing.
+     */
+    private final PrintWriter writer;
+
+    /**
+     * Creates a new {@code SuppressionFileGenerator} instance.
+     * Sets the output to a defined stream.
+     */
+    public XpathFileGeneratorAuditListener() {
+        writer = new PrintWriter(new OutputStreamWriter(System.out, StandardCharsets.UTF_8));
+
+        writer.println("<?xml version=\"1.0\" encoding=\"UTF-8\"?>");
+        writer.println("<!DOCTYPE suppressions PUBLIC");
+        writer.println("    \"-//Puppy Crawl//DTD Suppressions Xpath Experimental 1.2//EN\"");
+        writer.println("    \"http://checkstyle.sourceforge.net/dtds"
+                + "/suppressions_1_2_xpath_experimental.dtd\">");
+    }
+
+    @Override
+    public void auditStarted(AuditEvent event) {
+        writer.println("<suppressions>");
+    }
+
+    @Override
+    public void auditFinished(AuditEvent event) {
+        writer.println("</suppressions>");
+        writer.flush();
+    }
+
+    @Override
+    public void fileStarted(AuditEvent event) {
+        // No code by default
+    }
+
+    @Override
+    public void fileFinished(AuditEvent event) {
+        // No code by default
+    }
+
+    @Override
+    public void addError(AuditEvent event) {
+        final String xpathQuery = findCorrespondingXpathQuery(event);
+        if (xpathQuery != null) {
+            final String checkName = event.getSourceName();
+            final String fileName = event.getFileName();
+            final File file = new File(fileName);
+
+            writer.println("<suppress-xpath");
+            writer.print("       files=\"");
+            writer.print(file.getName());
+            writer.println(QUOTE_CHAR);
+
+            writer.print("       checks=\"");
+            writer.print(checkName);
+            writer.println(QUOTE_CHAR);
+
+            writer.print("       query=\"");
+            writer.print(xpathQuery);
+
+            writer.println("\"/>");
+        }
+    }
+
+    @Override
+    public void addException(AuditEvent event, Throwable throwable) {
+        throw new UnsupportedOperationException("Operation is not supported");
+    }
+
+    /**
+     * Returns xpath query corresponding to {@code TreeWalkerAuditEvent} object which points to
+     * the same AST element as specified {@code AuditEvent} object.
+     * @param event the {@code AuditEvent} object.
+     * @return returns corresponding xpath query
+     */
+    private String findCorrespondingXpathQuery(AuditEvent event) {
+        String result = null;
+        for (int i = 0; i < XpathFileGeneratorAstFilter.getEvents().size(); i++) {
```

Comment:
I think you can use "foreach" here, and do not mess with `i`

Reply:
I need `i` since I extract corresponding xpath query from another list

Reply:
ok, Now I see it.